vng/vng_test.go

```diff
@@ -0,0 +1,351 @@
+			builder.BeginContainer()
+			builder.Append(zed.EncodeInt(int64(tag)))
+			genValue(b, context, typ.Types[tag], builder)
+			builder.EndContainer()
+		default:
+			panic("Unreachable")
+		}
+	}
+}
+
+func genTypes(b *bytes.Reader, context *zed.Context, depth int) []zed.Type {
+	types := make([]zed.Type, 0)
+	for len(types) == 0 || genByte(b) != 0 {
+		types = append(types, genType(b, context, depth))
+	}
+	return types
+}
+
+func genType(b *bytes.Reader, context *zed.Context, depth int) zed.Type {
+	if depth < 0 || genByte(b)%2 == 0 {
+		switch genByte(b) % 19 {
+		case 0:
+			return zed.TypeUint8
+		case 1:
+			return zed.TypeUint16
+		case 2:
+			return zed.TypeUint32
+		case 3:
+			return zed.TypeUint64
+		case 4:
+			return zed.TypeInt8
+		case 5:
+			return zed.TypeInt16
+		case 6:
+			return zed.TypeInt32
+		case 7:
+			return zed.TypeInt64
+		case 8:
+			return zed.TypeDuration
+		case 9:
+			return zed.TypeTime
+		case 10:
+			// TODO Find a way to convert u16 to float16.
+			//return zed.TypeFloat16
+			return zed.TypeNull
+		case 11:
+			return zed.TypeFloat32
+		case 12:
+			return zed.TypeBool
+		case 13:
+			return zed.TypeBytes
+		case 14:
+			return zed.TypeString
+		case 15:
+			return zed.TypeIP
+		case 16:
+			return zed.TypeNet
+		case 17:
+			return zed.TypeType
+		case 18:
+			return zed.TypeNull
+		default:
+			panic("Unreachable")
+		}
+	} else {
+		depth := depth - 1
+		switch genByte(b) % 5 {
+		case 0:
+			fieldTypes := genTypes(b, context, depth)
+			fields := make([]zed.Field, len(fieldTypes))
+			for i, fieldType := range fieldTypes {
+				fields[i] = zed.Field{
+					Name: fmt.Sprint(i),
+					Type: fieldType,
+				}
+			}
+			typ, err := context.LookupTypeRecord(fields)
+			if err != nil {
+				panic(err)
+			}
+			return typ
+		case 1:
+			elem := genType(b, context, depth)
+			return context.LookupTypeArray(elem)
+		case 2:
+			key := genType(b, context, depth)
+			value := genType(b, context, depth)
+			return context.LookupTypeMap(key, value)
+		case 3:
+			elem := genType(b, context, depth)
+			return context.LookupTypeSet(elem)
+		case 4:
+			types := genTypes(b, context, depth)
+			// TODO There are some weird corners around unions that contain null or duplicate types eg
+			// vng_test.go:107: comparing: in[0]=null((null,null)) vs out[0]=null((null,null))
+			// vng_test.go:112: values have different zng bytes: [1 0] vs [2 2 0]
+			unionTypes := make([]zed.Type, 0)
+			for _, typ := range types {
+				skip := false
+				if typ == zed.TypeNull {
+					skip = true
+				}
+				for _, unionType := range unionTypes {
+					if typ == unionType {
+						skip = true
+					}
+				}
+				if !skip {
+					unionTypes = append(unionTypes, typ)
+				}
+			}
+			if len(unionTypes) == 0 {
+				return zed.TypeNull
+			}
+			return context.LookupTypeUnion(unionTypes)
+		default:
+			panic("Unreachable")
+		}
+	}
+}
+
+func genByte(b *bytes.Reader) byte {
+	// If we're out of bytes, return 0.
+	byte, _ := b.ReadByte()
```

Comment:
Maybe this since io.EOF is the only error expected?
```suggestion
	byte, err := b.ReadByte()
	if err != nil && !errors.Is(err, io.EOF) {
		panic(err)
	}	
```